Update shutter behavior and LF acquisition channel order

ieivanov opened the pull request:
This PR fixes mehta-lab/waveorder#361. It also starts to address ~#110~ and mehta-lab/waveorder#346 

To fully fix mehta-lab/waveorder#346 we also need to:
* ask the user to manually close and open the shutter if an auto shutter is not available
* provide the option to manually enter a black level  - either in the GUI or the CLI
@talonchandler could you please help me with those? 

~Towards addressing mehta-lab/waveorder#362 I've added a check to ensure that all LF channel exposures are the same. If they are not, all exposures are set to the exposure for State0. That's not a complete solution, but will prevent acquiring corrupted data.~

ieivanov committed:
fix shutter bugs

ieivanov committed:
fix load_calibration bugs

ieivanov committed:
don't update calibration metadata

ieivanov committed:
change pol data acquisition order

ieivanov commented on recOrder/calib/Calibration.py:
```diff
@@ -639,13 +644,41 @@ def opt_I135(self, lca_bound, lcb_bound):
+        self._shutter_state = self.mmc.getShutterOpen()
+
+        if not self.shutter_device:
+            # ask user to close shutter
+            pass
```
@talonchandler would appreciate your help here

ieivanov commented on recOrder/calib/Calibration.py:
```diff
@@ -639,13 +644,41 @@ def opt_I135(self, lca_bound, lcb_bound):
+
+    def open_shutter(self):
+        if not self.shutter_device:
+            # ask user to open shutter
+            pass
```
and here

ieivanov commented on recOrder/calib/Calibration.py:
```diff
@@ -639,13 +644,41 @@ def opt_I135(self, lca_bound, lcb_bound):
-        self.mmc.setShutterOpen(False)
+        """
+        if not self.shutter_device:
+            # ask user to open shutter
+            pass
```
and here too :)

ieivanov commented on recOrder/plugin/widget/main_widget.py:
```diff
@@ -1799,21 +1799,25 @@ def handle_fluor_reconstructor_update(self, value):
 
-        with open(self.last_calib_meta_file, 'w') as file:
-            json.dump(current_json, file, indent=1)
+        pass
```
I've realized that the acquisition works try to do a weird update of the calibration metadata. I think the intent is to update the microscope parameters (such as NA or magnification) in the metadata file. However, we currently (at least in most of my use cases) don't enter such microscope parameters during LC calibration, and they are also not written at that time - `write_metadata` is typically called without the optional `microscope_params` argument. So this function will have no effect, and in fact it throws an error since we changed the `Microscope Parameters` key to `Microscope parameters`. 

The only time we currently write these microscope parameters is during background capture: https://github.com/mehta-lab/recOrder/blob/1318f2d50d9cb9057822bdb2e6016ec79ac05f3a/recOrder/plugin/workers/calibration_workers.py#L321

I think it's good to save some of these parameters (such as magnification, n_objective_media, objective_NA, condenser_NA, swing, pixel_size) in the calibration metadata. Parameters such as pad_z or bg_correction_option belong to the dataset metadata, not the calibration metadata.

I welcome help and input on this issue

talonchandler replied:
Great, thanks for the context here. I'll think more about what parameters should be saved alongside the calibration metadata. I'll probably break it off into a separate issue/PR since there's a bit of work to do here.

ieivanov commented on recOrder/plugin/workers/acquisition_workers.py:
```diff
@@ -878,6 +878,22 @@ def work(self):
+
+            channel_exposures = np.array(channel_exposures)
+            if not np.all(channel_exposures == channel_exposures[0]):
+                # TODO: warn user that not all channels exposures were the same
```
@talonchandler could you help me with a warning message here? I think a warning on the CLI only is enough, right?

talonchandler replied:
I think a warning at the command line is enough. I also need to make sure that the metadata is written correctly, but that should be easy.

talonchandler commented:
This is excellent, Ivan, thanks so much. I'll definitely be able to take this over the finish line. 

I really like your "warning + default to State0" solution for #110---good call. 

In the interests of onboarding Ziwen, I'm thinking of testing and completing the fixes for #110, #111, and #142 on three different branches off of this branch, merge them into this single PR, then merge this into `main`. Sound okay? Thanks for giving us a major headstart!

talonchandler committed:
Revert exposure check --- move to separate PR.

talonchandler commented:
I've moved the exposure check to a different PR #162 since the implementation needs work, and this PR will benefit from narrowed scope.

talonchandler committed:
Revert "don't update calibration metadata"

This reverts commit 55caabee83436989311eaa3555ae1bc40a270a24.

talonchandler committed:
Remove orphan `Calibration/run_calibration` code

No references to this code
All functionality was moved to `calibration_workers.py` and/or `main_widget.py`

talonchandler committed:
Correct shutter device check

talonchandler committed:
Add breaks for manually opening and closing the shutter

talonchandler committed:
Close shutter and calc black level in single function

talonchandler committed:
Keep shutter open for multi-slice, multi-channel MDAs

talonchandler committed:
Merge branch 'main' into bugfix-shutter-channelOrder

talonchandler committed:
Move napari to a requirement for better testing

talonchandler committed:
Merge branch 'bugfix-shutter-channelOrder' of https://github.com/mehta-lab/recOrder into bugfix-shutter-channelOrder

talonchandler committed:
Drop support for 3.7

talonchandler committed:
Update docs

talonchandler commented:
This PR is on the larger side. It includes:

**Features**
- keeping the shutter open during multi-channel and multi-slice acquisitions
- options for users to calibrate when a shutter isn't available
  - if a shutter isn't available, a warning appears prompting an input on the command line
  - the user can enter a number that will be used as a black level, or close the shutter manually and let recOrder estimate the black level
- slice-first acquisition instead of channel-first acquisition

**Non-features**
- To allow for testing of import statements, I have made `napari[all]` a requirement (and updated the docs accordingly).
- Dropped support for `python 3.7` because `napari` has also dropped support. 

@ziw-liu much of this depends on hardware, but I'd still appreciate a quick second set of eyes on the code before merging. Thanks!

ziw-liu commented on .github/workflows/test_and_deploy.yml:
```diff
@@ -21,7 +21,7 @@ jobs:
       matrix:
         platform: [ubuntu-latest, windows-latest, macos-latest]
-        python-version: [3.7, 3.8, 3.9]
+        python-version: [3.8, 3.9]
```
The `python-version` specified here is not consistent with `pytests.yml`, namely lacking Python 3.10. Are we supporting for 3.10?

talonchandler replied:
Good catch! Fixed.

ziw-liu commented on recOrder/calib/Calibration.py:
```diff
@@ -145,6 +146,11 @@ def __init__(self, mmc, mm, group='Channel', lc_control_mode='MM-Retardance', in
         self.inst_mat = None
 
+        # Shutter
+        self.shutter_device = self.mmc.getShutterDevice()
```
What is the expected return value if no shutter is available to the MMCore?

ziw-liu replied:
From `QLIPP_Calibration.open_shutter` it seems like an empty string is expected. Need testing with hardware.

ziw-liu commented on recOrder/calib/Calibration.py:
```diff
@@ -639,13 +645,41 @@ def opt_I135(self, lca_bound, lcb_bound):
+
+        if self.shutter_device == '':  # no shutter
+            show_warning('No shutter found. Please follow the command-line instructions...')
+            in_string = input('recOrder could not find an automatic shutter configured through MicroManager.\n'
```
How about:

```python
shutter_warning_msg = """
recOrder could not find an automatic shutter configured through Micro-Manager.
>>> If you would like manually enter the black level, enter an integer or float and press <Enter>
>>> If you would like to estimate the black level, please close the shutter and press <Enter> 
"""

in_string = input(shutter_warning_msg)
```

talonchandler replied:
Much nicer! Thanks.

ziw-liu commented on recOrder/calib/Calibration.py:
```diff
@@ -655,15 +689,8 @@ def calc_blacklevel(self):
         self.I_Black = blacklevel
-
-        return blacklevel
+        return
```
No need for explicit `return` here.

talonchandler replied:
Nice!

ziw-liu commented on recOrder/plugin/workers/acquisition_workers.py:
```diff
@@ -872,9 +873,16 @@ def work(self):
             self._check_abort()
+
+            # set acquisition order to channel-first
+            settings['slicesFirst'] = False
```
Sorry that my [#162](https://github.com/mehta-lab/recOrder/pull/162) code caused this conflict, but you will need `self.settings` here.

talonchandler replied:
Not your fault! Great catch.

ziw-liu commented on recOrder/calib/Calibration.py:
```diff
@@ -639,13 +645,41 @@ def opt_I135(self, lca_bound, lcb_bound):
-        self.mmc.setShutterOpen(False)
+        """
+        if self.shutter_device == '': # no shutter
+            input('Please reset the shutter to its original state and press <Enter>')
```
Do we want to direct the user back to the GUI here? (e.g. `logging.info("This is the end of the commandline instructions.")`)

talonchandler replied:
Great idea!

ziw-liu commented on tox.ini:
```diff
@@ -1,12 +1,11 @@
 # For more information about tox, see https://tox.readthedocs.io/en/latest/
 [tox]
-envlist = py{37, 38,39}-{linux,macos,windows}
+envlist = py{38,39}-{linux,macos,windows}
```
Also no Python 3.10 here.

talonchandler replied:
Fixed!

talonchandler committed:
Test against 3.10

talonchandler committed:
Improved shutter warning msg.

talonchandler committed:
Remove unnecessary return

talonchandler committed:
Tell user to return to napari window.

talonchandler committed:
settings -> self.settings

ieivanov commented:
Quick note: I see that you've removed `run_calibration`, `run_5state_calibration`, and `run_4state_calibration` from `calib.Calibration`. These function used to let users run LC calibration outside of the `recOrder` widget (e.g. in a script). I think that's OK, as this is not a commonly used mode of `recOrder` right now, and we don't test and debug it.

talonchandler commented:
Thanks for the context @ieivanov. 

I'm totally in favor of script-based calibration if it's useful. For the sake of maintenance/updates, it will be great to have a single function that the GUI/scripts/others can call. I think we can revisit these design choices when we address #120.
